## EpiMethEx (Epigenetic Methylation and Expression)

### HOW-TO execute the script

1. Download the following dataset:

 * SKCM gene expression (pancan normalized) https://genome-cancer.ucsc.edu/download/public/TCGA_SKCM_exp_HiSeqV2_PANCAN-2015-02-24.tgz

 * SKCM DNA methylation Methylation450k https://genome-cancer.ucsc.edu/download/public/TCGA_SKCM_hMethyl450-2015-02-24.tgz 

 * Probset https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?mode=raw&is_datatable=true&acc=GPL13534&id=11288&db=GeoDb_blob92

2. Within the file **Script.R** you must set the following variables: 

  * `nameFolderDest` it's the absolute directory path where will be created within the `nameFD` of every execution.
  
  * `nameFD` it's the directory name which will contain the excel file of every range execution.
  
  * `PathDatasetPancan` absolute path of dataset **genomicMatrix** within the directory **TCGA_SKCM_exp_HiSeqV2_PANCAN-2015-02-24**
  
  * `PathDatasetProbe` absolute path of dataset **GPL13534-11288.txt** within the directory **probe**
  
  * `PathDatasetMethylation` absolute path of dataset **genomicMatrix** within the directory **TCGA_SKCM_hMethyl450-2015-02-24**
  
3.  Inside the instruction `dfPancan <-fread(PathDatasetPancan,header = T,sep = "\t")[1:10, ]` within square brackets change the range genes

### HOW-TO (DATA FILTERING)

Filtering data R script needs to be applied to all the tables generated by **Script.R**.

1. Set `nameFolderDest` variable with the path where you would save the filtered table
2. Change, eventually, the default value of the variables `Pvalue, FoldChange, etc..`
3. Execute the Script

### TESTING (System Configuration)
* OS: CentOS 7 - 64 bit
* CPU: 8 Cores
* RAM: 32GB
* R: 3.4.3
* MAX range of genes: 1000
