## EpiMethEx (Epigenetic Methylation and Expression)

### HOW-TO execute the script

1. Download the following dataset:

 * SKCM gene expression (pancan normalized) https://genome-cancer.ucsc.edu/download/public/TCGA_SKCM_exp_HiSeqV2_PANCAN-2015-02-24.tgz

 * SKCM DNA methylation Methylation450k https://genome-cancer.ucsc.edu/download/public/TCGA_SKCM_hMethyl450-2015-02-24.tgz 

 * Probset https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?mode=raw&is_datatable=true&acc=GPL13534&id=11288&db=GeoDb_blob92

2. Within the file **Script.R** you must set the following variables: 

  * `nameFolderDest` it's the absolute directory path where will be created within the `nameFD` of every execution.
  
  * `nameFD` it's the directory name which will contain the excel file of every execution.
  
  * `PathDatasetPancan` absolute path of dataset **genomicMatrix** within the directroy **TCGA_SKCM_exp_HiSeqV2_PANCAN-2015-02-24**
  
  * `PathDatasetProbe` absolute path of dataset **GPL13534-11288.txt** within the direcotry **probe**
  
  * `PathDatasetMethylation`path assoluto del dataset **genomicMatrix** all'interno della cartella **TCGA_SKCM_hMethyl450-2015-02-24**
  
3.  Inside the instruction `dfPancan <-fread(PathDatasetPancan,header = T,sep = "\t")[1:10, ]` within square brackets change the range genes

**_MAX range of genes: 1000 with PC configuration: RAM: 32GB CORES: 8_**



### HOW-TO (DATA FILTERING)

Filtering data R script needs to be applied to all the tables generated by **Script.R** code described in previous section. To select the table to apply filtering code to, user needs to execute the "read_excel(file.choose())" at the beginning of the "methylationsFilter.R" script. This script consists of six different filter types, and each one of these filters must be applied on a specific field of each table:

1. The first filter is applied on the *MedianUP*, *MedianMID* and *MedianDown* columns and extracts all values satisfying the conditions expressed in the code in order to identify maximum and minimum values among all medians.
2. The second filter is applied on the *ASS(bd_UPvsMID)*, *ASS(bd_UPvsDOWN)* and *ASS(bd_MIDvsDOWN)* columns of each table and extracts all data higher or equal to a specific fixed value (X=0.1).
3. The third filter is applied on the *pvalue_UPvsMID*, *pvalue_UPvsDOWN* and *pvalue_MIDvsDOWN* columns of each table and extracts all data higher or equal to a specific fixed value (X=0.01).
4. The fourth filter is applied on the *fc_UPvsMID.gene.*, *fc_UPvsDOWN.gene.* and *fc_MIDvsDOWN.gene.* columns of each table and extracts all data higher or equal to a specific fixed value (X=2).
5. The fifth filter is applied on the *pvalue_UPvsMID.gene.*, *pvalue_UPvsDOWN.gene.* and *pvalue_MIDvsDOWN.gene.* columns of each table and extracts all data higher or equal to a specific fixed value (X=0.01).
6. The sixth filter is applied on the *pvalue_pearson_correlation* column of each table and extracts all data higher or equal to a specific fixed value (X=0.05).

All six filters need to be sequentially applied on all tables in this specific order to extract only all data satisfying all conditions at once.
